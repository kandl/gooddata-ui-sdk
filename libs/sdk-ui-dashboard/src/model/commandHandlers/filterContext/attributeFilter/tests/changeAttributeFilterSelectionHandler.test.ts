// (C) 2021-2022 GoodData Corporation
import { beforeEach, describe, it, expect } from "vitest";
import { changeAttributeFilterSelection } from "../../../../commands/index.js";
import { DashboardTester, preloadedTesterFactory } from "../../../../tests/DashboardTester.js";
import { selectFilterContextAttributeFilters } from "../../../../store/filterContext/filterContextSelectors.js";
import { SimpleDashboardIdentifier } from "../../../../tests/fixtures/SimpleDashboard.fixtures.js";
import { TestCorrelation } from "../../../../tests/fixtures/Dashboard.fixtures.js";

describe("changeAttributeFilterSelectionHandler.test", () => {
    let Tester: DashboardTester;

    beforeEach(async () => {
        await preloadedTesterFactory((tester) => {
            Tester = tester;
        }, SimpleDashboardIdentifier);
    });

    it("should emit the appropriate events for changed attribute filter selection", async () => {
        const firstFilterLocalId = selectFilterContextAttributeFilters(Tester.state())[0].attributeFilter
            .localIdentifier!;

        Tester.dispatch(
            changeAttributeFilterSelection(
                firstFilterLocalId,
                { uris: ["testing/uri"] },
                "NOT_IN",
                TestCorrelation,
            ),
        );

        await Tester.waitFor("GDC.DASH/EVT.FILTER_CONTEXT.CHANGED");

        expect(Tester.emittedEventsDigest()).toMatchSnapshot();
    });

    it("should set the attribute selection in state on changed attribute filter selection", async () => {
        const firstFilterLocalId = selectFilterContextAttributeFilters(Tester.state())[0].attributeFilter
            .localIdentifier!;

        Tester.dispatch(
            changeAttributeFilterSelection(
                firstFilterLocalId,
                { uris: ["testing/uri"] },
                "NOT_IN",
                TestCorrelation,
            ),
        );

        await Tester.waitFor("GDC.DASH/EVT.FILTER_CONTEXT.CHANGED");

        expect(selectFilterContextAttributeFilters(Tester.state())[0]).toMatchSnapshot({
            attributeFilter: {
                localIdentifier: expect.any(String),
            },
        });
    });

    it("should emit the appropriate events when trying to change a non-existent attribute filter", async () => {
        Tester.dispatch(
            changeAttributeFilterSelection(
                "NON EXISTENT LOCAL ID",
                { uris: ["testing/uri"] },
                "NOT_IN",
                TestCorrelation,
            ),
        );

        await Tester.waitFor("GDC.DASH/EVT.COMMAND.FAILED");

        expect(Tester.emittedEventsDigest()).toMatchSnapshot();
    });

    it("should NOT alter the attribute filter state when trying to change a non-existent attribute filter", async () => {
        const originalFilters = selectFilterContextAttributeFilters(Tester.state());

        Tester.dispatch(
            changeAttributeFilterSelection(
                "NON EXISTENT LOCAL ID",
                { uris: ["testing/uri"] },
                "NOT_IN",
                TestCorrelation,
            ),
        );

        await Tester.waitFor("GDC.DASH/EVT.COMMAND.FAILED");

        expect(selectFilterContextAttributeFilters(Tester.state())).toEqual(originalFilters);
    });
});
