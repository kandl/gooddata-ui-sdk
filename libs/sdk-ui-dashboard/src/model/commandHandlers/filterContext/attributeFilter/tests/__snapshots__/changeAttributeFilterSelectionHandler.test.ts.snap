// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`changeAttributeFilterSelectionHandler.test > should emit the appropriate events for changed attribute filter selection 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.FILTER_CONTEXT.ATTRIBUTE_FILTER.CHANGE_SELECTION",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.SELECTION_CHANGED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.CHANGED",
  },
]
`;

exports[`changeAttributeFilterSelectionHandler.test > should emit the appropriate events when trying to change a non-existent attribute filter 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.FILTER_CONTEXT.ATTRIBUTE_FILTER.CHANGE_SELECTION",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.FAILED",
  },
]
`;

exports[`changeAttributeFilterSelectionHandler.test > should set the attribute selection in state on changed attribute filter selection 1`] = `
{
  "attributeFilter": {
    "attributeElements": {
      "uris": [
        "testing/uri",
      ],
    },
    "displayForm": {
      "uri": "/gdc/md/referenceworkspace/obj/1089",
    },
    "filterElementsBy": [],
    "localIdentifier": Any<String>,
    "negativeSelection": true,
  },
}
`;
